File: cloud/src/modules/workspace/types/knowledge.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeDocument {
    pub id: String,
    pub workspace_id: String,
    pub title: String,
    pub content: String,
    pub tags: Vec<String>,
    pub parse_status: String,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeEntity {
    pub id: String,
    pub workspace_id: String,
    pub source_document_id: String,
    pub entity_type: String,
    pub name: String,
    pub description: Option<String>,
    pub properties: serde_json::Value,
    pub tags: Vec<String>,
    pub file_ids: Vec<String>,
    pub device_id: Option<String>,
    pub confidence: f32,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeRelation {
    pub id: String,
    pub workspace_id: String,
    pub source_entity_id: String,
    pub target_entity_id: String,
    pub relation_type: String,
    pub properties: serde_json::Value,
    pub confidence: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeParseJob {
    pub id: String,
    pub document_id: String,
    pub status: String,
    pub error_message: Option<String>,
    pub result_summary: Option<ParseResultSummary>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParseResultSummary {
    pub entity_count: usize,
    pub relation_count: usize,
    pub diff: Option<ParseDiff>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParseDiff {
    pub added: usize,
    pub removed: usize,
    pub modified: usize,
}

// Request types
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateKnowledgeDocumentRequest {
    pub title: String,
    pub content: String,
    pub tags: Option<Vec<String>>,
    pub file_ids: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateKnowledgeDocumentRequest {
    pub title: Option<String>,
    pub content: Option<String>,
    pub tags: Option<Vec<String>>,
    pub file_ids: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateKnowledgeEntityRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub entity_type: Option<String>,
    pub properties: Option<serde_json::Value>,
    pub tags: Option<Vec<String>>,
    pub device_id: Option<Option<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PreviewParseRequest {
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PreviewParseResponse {
    pub entities: Vec<KnowledgeEntity>,
    pub relations: Vec<KnowledgeRelation>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeSearchResult {
    pub entity: KnowledgeEntity,
    pub relations: Vec<KnowledgeRelation>,
    pub source_snippet: String,
    pub relevance: f32,
}

// Query params
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeDocumentListParams {
    pub q: Option<String>,
    pub tags: Option<String>,
    pub status: Option<String>,
    pub page: Option<i64>,
    pub page_size: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeEntityListParams {
    pub entity_type: Option<String>,
    pub tags: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KnowledgeSearchParams {
    pub q: String,
    pub entity_type: Option<String>,
    pub tags: Option<String>,
    pub limit: Option<i64>,
}

File: web/src/api/knowledge.ts
/**
 * Knowledge Graph API
 */

import { apiGet, apiPost, apiPut, apiDelete, apiUpload } from './client.js';

export interface KnowledgeDocument {
  id: string;
  workspaceId: string;
  resourceType: string;
  name: string;
  description: string | null;
  content: string | null;
  filePath: string;
  fileSize: number | null;
  tags: string[];
  metadata: string | null;
  parseStatus: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface KnowledgeEntity {
  id: string;
  workspaceId: string;
  sourceDocumentId: string;
  entityType: string;
  name: string;
  description: string | null;
  properties: Record<string, unknown>;
  tags: string[];
  fileIds: string[];
  deviceId: string | null;
  confidence: number;
  createdAt: string;
  updatedAt: string;
}

export interface KnowledgeRelation {
  id: string;
  workspaceId: string;
  sourceEntityId: string;
  targetEntityId: string;
  relationType: string;
  properties: Record<string, unknown>;
  confidence: number;
}

export interface KnowledgeParseJob {
  id: string;
  documentId: string;
  status: 'pending' | 'running' | 'completed' | 'failed';
  errorMessage: string | null;
  resultSummary: ParseResultSummary | null;
  createdAt: string;
  updatedAt: string;
}

export interface ParseResultSummary {
  entityCount: number;
  relationCount: number;
  diff: ParseDiff | null;
}

export interface ParseDiff {
  added: number;
  removed: number;
  modified: number;
}

export interface CreateKnowledgeDocumentRequest {
  title: string;
  content: string;
  tags?: string[];
  fileIds?: string[];
}

export interface UpdateKnowledgeDocumentRequest {
  title?: string;
  content?: string;
  tags?: string[];
}

export interface PreviewParseRequest {
  content: string;
}

export interface PreviewParseResponse {
  entities: KnowledgeEntity[];
  relations: KnowledgeRelation[];
}

export interface KnowledgeSearchResult {
  entity: KnowledgeEntity;
  relations: KnowledgeRelation[];
  sourceSnippet: string;
  relevance: number;
}

export interface DocumentListResponse {
  data: KnowledgeDocument[];
  pagination: {
    page: number;
    pageSize: number;
    totalPages: number;
    totalCount: number;
  };
}

function getWorkspaceId(): string | null {
  if (typeof window === 'undefined') return null;
  return localStorage.getItem('workspace-id') || sessionStorage.getItem('workspace-id');
}

export const knowledgeApi = {
  // Documents
  async listDocuments(params?: { q?: string; tags?: string; status?: string; page?: number; pageSize?: number }) {
    const wsId = getWorkspaceId();
    if (!wsId) throw new Error('No workspace selected');
    return apiGet<DocumentListResponse>(`/workspaces/${wsId}/knowledge/documents`, params as Record<string, any>);
  },

  async createDocument(data: CreateKnowledgeDocumentRequest) {
    const wsId = getWorkspaceId();
    if (!wsId) throw new Error('No workspace selected');
    return apiPost<KnowledgeDocument>(`/workspaces/${wsId}/knowledge/documents`, data);
  },

  async getDocument(documentId: string) {
    const wsId = getWorkspaceId();
    if (!wsId) throw new Error('No workspace selected');
    return apiGet<KnowledgeDocument>(`/workspaces/${wsId}/knowledge/documents/${documentId}`);
  },

  async updateDocument(documentId: string, data: UpdateKnowledgeDocumentRequest) {
    const wsId = getWorkspaceId();
    if (!wsId) throw new Error('No workspace selected');
    return apiPut<KnowledgeDocument>(`/workspaces/${wsId}/knowledge/documents/${documentId}`, data);
  },

  async deleteDocument(documentId: string) {
    const wsId = getWorkspaceId();
    if (!wsId) throw new Error('No workspace selected');
    return apiDelete<void>(`/workspaces/${wsId}/knowledge/documents/${documentId}`);
  },

  // Parse
  async triggerParse(documentId: string) {
    const wsId = getWorkspaceId();
    if (!wsId) throw new Error('No workspace selected');
    return apiPost<{ parseId: string }>(`/workspaces/${wsId}/knowledge/documents/${documentId}/parse`);
  },

  async previewParse(documentId: string, data: PreviewParseRequest) {
    const wsId = getWorkspaceId();
    if (!wsId) throw new Error('No workspace selected');
    return apiPost<PreviewParseResponse>(`/workspaces/${wsId}/knowledge/documents/${documentId}/preview`, data);
  },

  async getParseJob(jobId: string) {
    const wsId = getWorkspaceId();
    if (!wsId) throw new Error('No workspace selected');
    return apiGet<KnowledgeParseJob>(`/workspaces/${wsId}/knowledge/parse/${jobId}`);
  },

  // Entities & Relations
  async listEntities(params?: { entityType?: string; tags?: string; documentId?: string }) {
    const wsId = getWorkspaceId();
    if (!wsId) throw new Error('No workspace selected');
    return apiGet<KnowledgeEntity[]>(`/workspaces/${wsId}/knowledge/entities`, params as Record<string, any>);
  },

  async updateEntity(entityId: string, data: Record<string, unknown>) {
    const wsId = getWorkspaceId();
    if (!wsId) throw new Error('No workspace selected');
    return apiPut<KnowledgeEntity>(`/workspaces/${wsId}/knowledge/entities/${entityId}`, data);
  },

  async listRelations() {
    const wsId = getWorkspaceId();
    if (!wsId) throw new Error('No workspace selected');
    return apiGet<KnowledgeRelation[]>(`/workspaces/${wsId}/knowledge/relations`);
  },

  // Search
  async searchKnowledge(params: { q: string; entityType?: string; tags?: string; limit?: number }) {
    const wsId = getWorkspaceId();
    if (!wsId) throw new Error('No workspace selected');
    return apiGet<KnowledgeSearchResult[]>(`/workspaces/${wsId}/knowledge/search`, params as Record<string, any>);
  },

  // Context (for debugging)
  async getContext() {
    const wsId = getWorkspaceId();
    if (!wsId) throw new Error('No workspace selected');
    return apiGet<string>(`/workspaces/${wsId}/knowledge/context`);
  },

  // File upload (reuses existing upload endpoint)
  async uploadFile(file: File, onProgress?: (pct: number) => void) {
    const wsId = getWorkspaceId();
    if (!wsId) throw new Error('No workspace selected');
    const formData = new FormData();
    formData.append('file', file);
    return apiUpload<{ filePath: string; fileSize: number }>(
      `/workspaces/${wsId}/resources/upload`,
      formData,
      onProgress,
    );
  },
};
